### DISCUSSION TITLE
Layout Reset / Groups / Pathless Routes

### DISCUSSION THREAD

**Discussion by rmarscher** [wakujs/waku#977] [General]
Have you considered allowing nested pages or layouts with fs router to "reset" so they don't include the parent layouts? I try to avoid that and don't currently need it, but the use case can arise.

Other routers ([SvelteKit](https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-group), [NextJS](https://nextjs.org/docs/app/building-your-application/routing/route-groups), [Remix](https://remix.run/docs/en/main/file-conventions/routes#nested-layouts-without-nested-urls), [Tanstack Router](https://tanstack.com/router/latest/docs/framework/react/guide/file-based-routing#pathless-routes)) have added groups or "pathless routes" to their fs routers as another way to solve similar use cases.    

Or revisit #272 to keep Waku's router from sprawling into these advanced use cases?

**Comment by dai-shi** [comment]
I'm not sure how we call it, but I would like to introduce something like pathless components. With that in mind, #949 is already designed to be flexible to support it. So, the next step is to design the createPages API, after v0.22.0. cc @tylersayshi 

I'm not sure about layout reset, but the same technique should be applied. You shouldn't use the parent layout if you ever want to reset it.

**Reply by tylersayshi** [reply]
Yeah, `createLayout` could acknowledge something like `path: '/(group)'` which would then allow us to solve for this.

There are probably other ideas for handling this with the api. This is the first that comes to mind though.

**Reply by rmarscher** [reply]
OK great. I agree that layout reset is not required if something like pathless components is available. Thank you :)

**Comment by dai-shi** [comment]
@tylersayshi 
Re: https://github.com/dai-shi/waku/discussions/977#discussioncomment-10991421

> There are probably other ideas for handling this with the api. This is the first that comes to mind though.

Yes, I think there should be a new API. I'd also like to consider supporting siblings.

> `path: '/(group)'`

I'm strictly against to this in createPages API. createPages API should be more explicit and in fs-router, it can be `(group).tsx`.

**Reply by tylersayshi** [reply]
hmm `createLayout` could take a list of paths that they apply to? e.g. `path: ['/about', '/contact']`

Is this closer to what you'd think? Or are there other thoughts?

**Reply by dai-shi** [reply]
A little different. I would like to make it more flexible.
```tsx
// `/foo/bar`
<TopLayout>
  <FooLayout>
    <StaticComponent />
    <DynamicComponent />
    <BarLayout>
      <BarPage>
```

**Reply by tylersayshi** [reply]
@dai-shi have you thought more about how we may support this more flexible style?

having groups and combining layouts/pages within those groups make sense to me, but I don't yet understand how this more flexible style of composing layouts/pages (and other components) could be possible with the fs-router projects.

**Reply by dai-shi** [reply]
On second thought, I actually think `(group)` is enough to support what I want.
```
./pages/foo/(static)/(dynamic)/bar/index.tsx
```

On createPages API, we should introduce a new `createGroup` api.
I'm not sure if we should allow layout in group.

**Reply by tylersayshi** [reply]
why would we disallow layouts in groups? That seems to be a big reason users would want to create groups in the first place; to group different sets of layouts.

**Reply by dai-shi** [reply]
I mean createGroup would be essentially the same as createLayout. I wasn't quite sure how the api was going to be. Maybe we can start discussing createPages api.

**Reply by tylersayshi** [reply]
If we follow how other frameworks do groups, they are not "the same as createLayout".

This screenshot from nextjs is helpful I think to illustrating how this concept works

<img width="463" alt="image" src="https://github.com/user-attachments/assets/38e75438-30ea-4351-be69-463074f250c0" />

route groups can have routes, layouts, nested routes, and nested layouts. From a quick scan of the other frameworks linked above by @rmarscher - this is a consistent pattern, so it seems good to follow to me.

**Reply by dai-shi** [reply]
Yeah, my confusion. Let's discuss fs-router, before createPages. 

To make this structure:
```jsx
// `/foo/bar`
<TopLayout>
  <FooLayout>
    <StaticComponent />
    <DynamicComponent />
    <BarLayout>
      <BarPage>
```

We need this file system:
```
- ./pages/_layout.tsx
- ./pages/foo/_layout.tsx
- ./pages/foo/(static)/_layout.tsx
- ./pages/foo/(static)/(dynamic)/_layout.tsx
- ./pages/foo/(static)/(dynamic)/bar/_layout.tsx
- ./pages/foo/(static)/(dynamic)/bar/index.tsx
```

`./pages/foo/(static)/_layout.tsx` would be like this for example:
```tsx
import { StaticComponent } from '...';

export default function StaticGroupLayout({ children }) {
  return (
    <>
      <StaticComponent />
      {children}
    </>
  );
  // Or, return [<StaticComponent />, children] ?
}

export const getConfig = async () => {
  return {
    render: 'static',
  } as const;
};
```

**Reply by tylersayshi** [reply]
If I follow correctly, you could also describe this as:

```tsx
<TopLayout>
  <FooLayout>
    <StaticLayout>
      <DynamicLayout>
        <BarLayout>
          <BarPage>
```

this would be what it looks like before knowing the contents of the layouts

**Reply by dai-shi** [reply]
Ohh, good point. So, I think my requirement is slightly different from `(group)`.

Let's separate them:
1. Support `(group)`
2. Support static component in dynamic layout https://github.com/dai-shi/waku/discussions/977#discussioncomment-10999003

What I wonder is at the createPages API level, the capability is similar or the same.

**Reply by tylersayshi** [reply]
ahh I see what you're looking for now. I agree both would be good and should be considered separately.

**Reply by dai-shi** [reply]
Hmm, in terms of "pathless" routes, I think two things are the same.

### Group routes

```
./pages/_layout.tsx
./pages/foo.tsx
./pages/(group)/_layout.tsx
./pages/(group)/bar.tsx
./pages/(group)/baz.tsx
```

### Static component in dynamic layout

```
./pages/_layout.tsx
./pages/foo/_layout.tsx
./pages/foo/(static).tsx
./pages/foo/(dynamic).tsx
./pages/foo/index.tsx
```

The question is how to order those components.

**Reply by rmarscher** [reply]
```
./pages/foo/_layout.tsx
./pages/foo/(static).tsx
./pages/foo/(dynamic).tsx
./pages/foo/index.tsx
```

This is interesting. Maybe the `(name).tsx` components could be passed as props to the layout and page render functions so they can decide where to place them in the component tree?

**Reply by dai-shi** [reply]
We've discussed internally and will solve these two requirements separately.
Here's the current idea:

### Group routes

Use `.../(group)/...` convention.

```
./pages/_layout.tsx
./pages/foo.tsx
./pages/(somegroup)/_layout.tsx
./pages/(somegroup)/bar.tsx
./pages/(somegroup)/baz.tsx
```

### Multiple components in a single page (Static component in dynamic layout)

Use `.../${n}.tsx` convention.

```
./pages/_layout.tsx
./pages/foo/_layout.tsx
./pages/foo/$0.tsx                    // can be static
./pages/foo/$1.tsx                    // can be dynamic
./pages/foo/$2.tsx
```

**Reply by dai-shi** [reply]
#1313 for the second one.

**Comment by dai-shi** [comment]
backlink: #1049

**Comment by rmarscher** [comment]
backlink: #561

**Comment by rmarscher** [comment]
I confirmed that this is now available in v0.21.21. Thanks @tylersayshi !!!

**Reply by dai-shi** [reply]
We plan to release the both features https://github.com/wakujs/waku/discussions/977#discussioncomment-11687612 in v0.23.

**Comment by dai-shi** [comment]
v0.23.0 has already been out.
